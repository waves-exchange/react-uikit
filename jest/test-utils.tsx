import React from 'react';
import { render, RenderResult, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';

type ProvidersWrapperProps = {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    theme: any;
};
const ProvidersWrapper: React.FC<ProvidersWrapperProps> = ({
    theme,
    children,
}) => {
    return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

interface CustomRenderOptions extends RenderOptions {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    theme: any;
}

type CustomRender = (
    Component: React.ReactElement,
    options: CustomRenderOptions
) => RenderResult;

const customRender: CustomRender = (Component, { theme, ...options }) => {
    return render(
        <ProvidersWrapper theme={theme}>{Component}</ProvidersWrapper>,
        options
    );
};

export * from '@testing-library/react';
export { customRender as render };
