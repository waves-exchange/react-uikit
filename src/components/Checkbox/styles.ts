import { IControlBoxStyles } from '../ControlBox/ControlBox';

export const defaultControlBoxStyles: IControlBoxStyles = {
    baseStyles: {
        userSelect: 'none' as 'none',
        border: '1px solid',
        color: 'white',
        borderColor: 'main.$500',
        backgroundColor: 'main.$800',
        borderRadius: '$4',
        transition: 'all 120ms, background-color 120ms, box-shadow 250ms',
        width: '18px',
        height: '18px',
    },
    _hover: {
        borderColor: 'main.$200',
    },
    _child: {
        opacity: 0,
        width: '100%',
        height: '100%',
    },
    _checkedAndChild: {
        opacity: 1,
    },
    _checked: {
        backgroundColor: 'primary.$300',
        borderColor: 'primary.$300',
        ':hover': {
            borderColor: 'main.$200',
        },
    },
    _checkedAndHover: {
        borderColor: 'primary.$300',
    },
    _checkedAndDisabled: {
        backgroundColor: 'primary.$900',
        borderColor: 'primary.$900',
    },
    _checkedAndDisabledHover: {
        borderColor: 'primary.$900',
    },
    _disabled: {
        borderColor: 'main.$500',
        backgroundColor: 'main.$1000',
    },
};
