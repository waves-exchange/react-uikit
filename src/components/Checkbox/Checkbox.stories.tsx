import React from 'react';
import { Flex } from '../Flex/Flex';
import { defaultTheme } from '../../themes/default';
import { ThemeProvider } from 'emotion-theming';
import { storiesOf } from '@storybook/react';
import { ControlBox } from '../ControlBox/ControlBox';
import { Checkbox } from './Checkbox';
import { defaultControlBoxStyles } from './styles';
import { mergeDeepRight } from 'ramda';
import { Icon } from '../Icon/Icon';
import { iconCheck } from '../../icons/check';

// TODO нужны тесты на чек иконки, на корректную работу label for + чекбокс

const stories = storiesOf('Checkbox', module);

stories.add('simple', () => {
    const cbStyles = {
        baseStyles: {
            marginRight: '16px',
        },
    };

    const customControlBoxStyles = mergeDeepRight(defaultControlBoxStyles, {
        baseStyles: {
            marginRight: '16px',
            borderColor: 'red',
            width: '20px',
            height: '20px',
        },
    });

    return (
        <ThemeProvider theme={defaultTheme}>
            <Flex
                p={16}
                bg="main.$800"
                flexDirection="column"
                justifyContent="center"
                color="white"
            >
                <Checkbox defaultChecked={false} controlBoxStyles={cbStyles}>
                    default unchecked
                </Checkbox>
                <Checkbox
                    defaultChecked={true}
                    pt={10}
                    controlBoxStyles={cbStyles}
                >
                    default checked
                </Checkbox>
                <Checkbox
                    defaultChecked={false}
                    isDisabled={true}
                    pt={10}
                    controlBoxStyles={cbStyles}
                >
                    disabled unchecked
                </Checkbox>
                <Checkbox
                    defaultChecked={true}
                    isDisabled={true}
                    pt={10}
                    controlBoxStyles={cbStyles}
                >
                    disabled checked
                </Checkbox>
                <Checkbox
                    defaultChecked={false}
                    pt={10}
                    controlBoxStyles={cbStyles}
                    controlBox={(): React.ReactNode => (
                        <ControlBox
                            sx={customControlBoxStyles.baseStyles}
                            _focus={customControlBoxStyles._focus}
                            _hover={customControlBoxStyles._hover}
                            _invalid={customControlBoxStyles._disabled}
                            _disabled={customControlBoxStyles._disabled}
                            _checked={customControlBoxStyles._checked}
                            _child={customControlBoxStyles._child}
                            _checkedAndChild={
                                customControlBoxStyles._checkedAndChild
                            }
                            _checkedAndDisabled={
                                customControlBoxStyles._checkedAndDisabled
                            }
                            _checkedAndDisabledHover={
                                customControlBoxStyles._checkedAndDisabledHover
                            }
                            _checkedAndFocus={
                                customControlBoxStyles._checkedAndFocus
                            }
                            _checkedAndHover={
                                customControlBoxStyles._checkedAndHover
                            }
                        >
                            <Icon icon={iconCheck} />
                        </ControlBox>
                    )}
                >
                    custom
                </Checkbox>
                <Checkbox
                    isReadOnly={false}
                    defaultChecked={true}
                    isChecked={false}
                    pt={10}
                    controlBoxStyles={cbStyles}
                >
                    disabled checked
                </Checkbox>
            </Flex>
        </ThemeProvider>
    );
});
