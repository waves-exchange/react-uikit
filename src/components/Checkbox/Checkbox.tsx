import React, { InputHTMLAttributes } from 'react';
import { Flex, TFlexProps } from '../Flex/Flex';
import { Icon } from '../Icon/Icon';
import { ControlBox, IControlBoxStyles } from '../ControlBox/ControlBox';
import { mergeDeepRight } from 'ramda';
import { defaultControlBoxStyles } from './styles';
import { iconCheck } from '../../assets/icons/check';

/**
 * Usage notes:
 * To customize either pass controlBoxStyles (see ControlBox component for the list of styles props)
 * or render prop ControlBox with your styles: () => <ControlBox {...styles} />
 * see storybook for examples
 */

interface CheckboxSpecificProps {
    controlBoxStyles?: IControlBoxStyles;
    controlBox?: () => React.ReactNode;
    isChecked?: boolean;
    isFullWidth?: boolean;
    isDisabled?: boolean;
    isReadOnly?: boolean;
}

export type ICheckboxProps = CheckboxSpecificProps &
    TFlexProps &
    InputHTMLAttributes<HTMLInputElement>;

export const Checkbox: React.FC<ICheckboxProps> = ({
    children,
    controlBox,
    controlBoxStyles = {},
    id,
    name,
    value,
    isChecked,
    defaultChecked,
    isFullWidth,
    isDisabled,
    isReadOnly,
    onChange,
    onBlur,
    onFocus,
    ...rest
}) => {
    const controlStyles = mergeDeepRight(
        defaultControlBoxStyles,
        controlBoxStyles
    ) as IControlBoxStyles;

    const { baseStyles, ...restControlStyles } = controlStyles;

    return (
        <Flex
            as="label"
            display="inline-flex"
            verticalAlign="top"
            alignItems="center"
            width={isFullWidth ? 'full' : undefined}
            cursor={isDisabled ? 'not-allowed' : 'pointer'}
            {...rest}
        >
            <input
                style={{
                    border: '0px',
                    clip: 'rect(0px, 0px, 0px, 0px)',
                    height: '1px',
                    width: '1px',
                    margin: '-1px',
                    padding: '0px',
                    overflow: 'hidden',
                    whiteSpace: 'nowrap',
                    position: 'absolute',
                }}
                id={id}
                type="checkbox"
                name={name}
                value={value}
                onChange={isReadOnly ? undefined : onChange}
                onBlur={onBlur}
                onFocus={onFocus}
                defaultChecked={isReadOnly ? undefined : defaultChecked}
                checked={
                    isReadOnly
                        ? Boolean(isChecked)
                        : defaultChecked
                        ? undefined
                        : isChecked
                }
                disabled={isDisabled}
                readOnly={isReadOnly}
            />
            {controlBox ? (
                controlBox()
            ) : (
                <ControlBox sx={baseStyles} {...restControlStyles}>
                    <Icon icon={iconCheck} />
                </ControlBox>
            )}
            {children}
        </Flex>
    );
};
