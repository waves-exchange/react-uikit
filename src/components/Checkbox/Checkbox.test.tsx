import { ICheckboxProps, controlBoxTestId, checkBoxTestId } from './Checkbox';

import React, { ReactNode, FC } from 'react';
import { render, fireEvent } from 'test-utils';
import { Checkbox, checkBoxInputTestId } from './Checkbox';

const defaultPropsFactory = (props: ICheckboxProps = {}): ICheckboxProps => {
    return {
        isFullWidth: false,
        isDisabled: false,
        isReadOnly: false,
        onChange: jest.fn(),
        onBlur: jest.fn(),
        onFocus: jest.fn(),
        id: 'id',
        ...props,
    };
};

describe('<Checkbox />', () => {
    const label = 'label';

    it('becomes checked on click', () => {
        const { getByLabelText, getByTestId } = render(
            <Checkbox {...defaultPropsFactory()}>{label}</Checkbox>
        );

        const inputElem = getByTestId(checkBoxInputTestId);

        expect(inputElem).not.toBeChecked();
        fireEvent.click(getByLabelText(label));
        expect(inputElem).toBeChecked();
    });

    it('renders control box with correct styles', () => {
        const { getByTestId } = render(
            <Checkbox {...defaultPropsFactory()}>{label}</Checkbox>
        );

        expect(getByTestId(controlBoxTestId)).toMatchSnapshot();
    });

    it('renders custom control box', () => {
        const customControlBoxText = 'test';

        const CustomControl: FC = () => <div>{customControlBoxText}</div>;

        const { getByText } = render(
            <Checkbox
                {...defaultPropsFactory()}
                controlBox={(): ReactNode => <CustomControl />}
            >
                {label}
            </Checkbox>
        );

        getByText(customControlBoxText);
    });

    it('forwards style props', () => {
        const marginTop = 40;

        const { getByTestId } = render(
            <Checkbox {...defaultPropsFactory()} mt={marginTop}>
                {label}
            </Checkbox>
        );

        expect(getByTestId(checkBoxTestId)).toHaveStyle(
            `margin-top: ${marginTop}px`
        );
    });

    describe('prop: "isFullWidth"', () => {
        it('width 100% if "true"', () => {
            const { getByTestId } = render(
                <Checkbox {...defaultPropsFactory({ isFullWidth: true })}>
                    {label}
                </Checkbox>
            );

            expect(getByTestId(checkBoxTestId)).toHaveStyleRule(
                'width',
                'full'
            );
        });

        it('no width if "false"', () => {
            const { getByTestId } = render(
                <Checkbox {...defaultPropsFactory({ isFullWidth: false })}>
                    {label}
                </Checkbox>
            );

            expect(getByTestId(checkBoxTestId)).not.toHaveStyleRule(
                'width',
                'full'
            );
        });
    });

    describe('prop: "isDisabled"', () => {
        it('cursor "not-allowed" if "true"', () => {
            const { getByTestId } = render(
                <Checkbox {...defaultPropsFactory({ isDisabled: true })}>
                    {label}
                </Checkbox>
            );

            expect(getByTestId(checkBoxTestId)).toHaveStyleRule(
                'cursor',
                'not-allowed'
            );
        });

        it('cursor "pointer" if "false"', () => {
            const { getByTestId } = render(
                <Checkbox {...defaultPropsFactory({ isDisabled: false })}>
                    {label}
                </Checkbox>
            );

            expect(getByTestId(checkBoxTestId)).toHaveStyleRule(
                'cursor',
                'pointer'
            );
        });
    });
    describe('prop: "onChange"', () => {
        it('calls if isReadOnly !== true', () => {
            const defaultProps = defaultPropsFactory({ isReadOnly: false });
            const { onChange } = defaultProps;

            const { getByLabelText } = render(
                <Checkbox {...defaultProps}>{label}</Checkbox>
            );

            fireEvent.click(getByLabelText(label));

            expect(onChange).toHaveBeenCalledTimes(1);
        });

        it("doesn't call if isReadOnly === true", () => {
            const defaultProps = defaultPropsFactory({ isReadOnly: true });
            const { onChange } = defaultProps;

            const { getByLabelText } = render(
                <Checkbox {...defaultProps}>{label}</Checkbox>
            );

            fireEvent.click(getByLabelText(label));

            expect(onChange).not.toHaveBeenCalled();
        });
    });

    describe('becomes checked when passed props', () => {
        it('isReadOnly === true && defaultChecked === true', () => {
            const defaultChecked = true;

            const { getByTestId } = render(
                <Checkbox
                    {...defaultPropsFactory({
                        isReadOnly: false,
                        defaultChecked,
                    })}
                >
                    {label}
                </Checkbox>
            );

            expect(getByTestId(checkBoxInputTestId)).toBeChecked();
        });

        it('isReadOnly === true && isChecked === true', () => {
            const isChecked = true;

            const { getByTestId } = render(
                <Checkbox
                    {...defaultPropsFactory({
                        isReadOnly: true,
                        isChecked,
                    })}
                >
                    {label}
                </Checkbox>
            );

            expect(getByTestId(checkBoxInputTestId)).toBeChecked();
        });

        it('defaultIsChecked === true && isChecked === false', () => {
            const { getByTestId } = render(
                <Checkbox
                    {...defaultPropsFactory({
                        isReadOnly: false,
                        defaultChecked: true,
                        isChecked: false,
                    })}
                >
                    {label}
                </Checkbox>
            );

            expect(getByTestId(checkBoxInputTestId)).toBeChecked();
        });
    });

    describe('becomes unchecked when passed props', () => {
        it('isReadOnly === true && defaultChecked === true', () => {
            const defaultChecked = true;

            const { getByTestId } = render(
                <Checkbox
                    {...defaultPropsFactory({
                        isReadOnly: true,
                        defaultChecked,
                    })}
                >
                    {label}
                </Checkbox>
            );

            expect(getByTestId(checkBoxInputTestId)).not.toBeChecked();
        });
    });

    it('calls "onFocus"', () => {
        const defaultProps = defaultPropsFactory();
        const { onFocus } = defaultProps;

        const { getByLabelText } = render(
            <Checkbox {...defaultProps}>{label}</Checkbox>
        );

        fireEvent.focus(getByLabelText(label));

        expect(onFocus).toHaveBeenCalled();
    });

    it('calls "onBlur"', () => {
        const defaultProps = defaultPropsFactory();
        const { onBlur } = defaultProps;

        const { getByLabelText } = render(
            <Checkbox {...defaultProps}>{label}</Checkbox>
        );

        const checkbox = getByLabelText(label);

        fireEvent.focus(checkbox);
        fireEvent.blur(checkbox);

        expect(onBlur).toHaveBeenCalled();
    });
});
