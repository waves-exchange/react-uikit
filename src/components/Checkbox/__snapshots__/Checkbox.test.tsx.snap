// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> renders control box with correct styles 1`] = `
.emotion-1 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: 1px solid;
  color: white;
  border-color: main.$600;
  background-color: basic.$900;
  border-radius: $4;
  -webkit-transition: all 120ms,background-color 120ms,box-shadow 250ms;
  transition: all 120ms,background-color 120ms,box-shadow 250ms;
  width: 18px;
  height: 18px;
}

input[type=checkbox]:hover:not(:disabled):not(:checked) + .emotion-1 {
  border-color: main.$200;
}

input[type=checkbox]:disabled + .emotion-1 {
  border-color: main.$500;
  background-color: main.$1000;
}

input[type=checkbox]:checked:disabled + .emotion-1,
input[type=checkbox][aria-checked=mixed]:disabled + .emotion-1 {
  background-color: primary.$900;
  border-color: primary.$900;
}

input[type=checkbox]:checked:hover:not(:disabled) + .emotion-1,
input[type=checkbox][aria-checked=mixed]:hover:not(:disabled) + .emotion-1 {
  border-color: primary.$300;
}

input[type=checkbox]:hover:disabled:checked + .emotion-1 {
  border-color: primary.$900;
}

.emotion-1 > * {
  opacity: 0;
  width: 100%;
  height: 100%;
}

input[type=checkbox]:checked + .emotion-1,
input[type=checkbox][aria-checked=mixed] + .emotion-1 {
  background-color: primary.$300;
  border-color: primary.$300;
}

input[type=checkbox]:checked + .emotion-1:hover,
input[type=checkbox][aria-checked=mixed] + .emotion-1:hover {
  border-color: main.$200;
}

input[type=checkbox]:checked + .emotion-1 > *,
input[type=checkbox][aria-checked=mixed] + .emotion-1 > * {
  opacity: 1;
}

.emotion-0 {
  box-sizing: border-box;
  color: currentColor;
  display: inline-block;
  vertical-align: middle;
  width: 14px;
  height: 14px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
}

.emotion-0:not(:root) {
  overflow: hidden;
}

<div
  aria-hidden="true"
  class="emotion-1"
  data-testid="control-box"
>
  <svg
    class="emotion-0"
    viewBox="0 0 14 14"
  >
    <g
      fill="currentColor"
      fill-rule="evenodd"
    >
      <path
        d="M10.315 3.5L5.56 8.176 3.68 6.333 2.5 7.492 5.56 10.5l5.94-5.842z"
        fill="currentColor"
      />
    </g>
  </svg>
</div>
`;
